jupyter-extension/jupyterlab_pachyderm/pfs_manager.py

```diff
@@ -469,12 +474,62 @@ def current_datum(self) -> dict:
             all_datums_received=True,
         )
 
-    def _get_fileinfo_path(self, fileinfo: pfs.FileInfo) -> Path:
+    def download(self):
+        if len(self._datum_list) == 0 or len(self._datum_list[self._datum_index].data) == 0:
+            raise ValueError("Attempting to download empty or unmounted datum")
+
+        download_dir = (
+            f"{self._DOWNLOAD_DIR}/datum-{datetime.datetime.now().isoformat()}"
+        )
+        try:
+            os.makedirs(download_dir, exist_ok=True)
+            for fileinfo in self._datum_list[self._datum_index].data:
+                path = self._get_download_path(
+                    download_dir=download_dir, fileinfo=fileinfo
+                )
+                os.makedirs(path.parent, exist_ok=True)
+                if fileinfo.file_type == pfs.FileType.FILE:
+                    # download individual file
+                    data = list(self._client.pfs.get_file(file=fileinfo.file))[0].value
+                    with open(path, "wb") as download_file:
+                        download_file.write(data)
+                elif fileinfo.file_type == pfs.FileType.DIR:
+                    # download tarball
+                    tar = self._client.pfs.pfs_tar_file(file=fileinfo.file)
+                    tar.extractall(path=path if fileinfo.file.path == "/" else path.parent)
+                else:
+                    raise TypeError(
+                        f"Attempting to download invalid file type {fileinfo.file_type}"
+                    )
+
+            for dir in os.listdir(PFS_MOUNT_DIR):
```

Comment:
would it be safer to iterate over `os.listdir(self._download_dir)` instead?

Reply:
If I'm not mistaken that doesn't accomplish the same thing; `/pfs/dir` in this instance would be the symlink under `/pfs` which we want removed. Removing the actual download is done below in the `shutil.rmtree` call.

Reply:
Oh so in most cases this is just unlinking a single directory (the single datum downloaded)?

Reply:
It would be just the single datum but not necessarily a single directory. For example, if we had a cross input of multiple repos, it would download files from each of the repos. Each symlink created would be for one repo, mirroring what we used to do with the mount-server when mounting a datum.